### PR TITLE
Kahan

### PR DESCRIPTION
**Pull request by j-mathe**
# This Pull request:

## Changes or fixes:

Replace the Sum and Mean methods by the Kahan Sum method to improve summation precision.

## Checklist:

- [ ] tested changes locally
- [ ] updated the docs (if necessary)

### CODE REVIEW

**Comment by eguiraud** [comment]
Thank you @j-mathe ! The failure in the `dataframe_misc` test in roottest looks real. Before/after this patch the output of the test is:

```
-Sum b2 init: 2471
-Sum dv init: 1491
+Sum b2 init: 2470
+Sum dv init: 1490
```

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx [hunk 1]
```diff
@@ -1142,26 +1148,45 @@ public:
...
    void Initialize() { /* noop */}
 
    void Finalize()
    {
-      for (auto &m : fSums)
-         *fResultSum += m;
+   ResultType sum(0) ;
+   ResultType compensation(0);
+   ResultType y(0);
+   ResultType t(0);
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1181]: all these `0`s should be `NeutralElement(m, 0)` for generality. Then we could discuss whether this generality still makes sense now that we are strongly assuming that what we are summing are numbers or arrays thereof, but that's for a second step, for now let's keep code consistent.

File: tree/dataframe/test/dataframe_simple.cxx [hunk 2]
```diff
@@ -726,12 +726,21 @@ TEST_P(RDFSimpleTests, StandardDeviationEmpty)
...
    EXPECT_DOUBLE_EQ(*stdDev, 0);
 }
-
+/*
 TEST(RDFSimpleTests, SumOfStrings)
 {
    auto df = RDataFrame(2).Define("str", []() -> std::string { return "bla"; });
    EXPECT_EQ(*df.Sum<std::string>("str"), "blabla");
 }
+*/
```
Comment (eguiraud) [inline tree/dataframe/test/dataframe_simple.cxx:736]: this commit should come before the previous, otherwise the previous commit won't compile

File: tree/dataframe/test/dataframe_simple.cxx [hunk 3]
```diff
@@ -726,12 +726,21 @@ TEST_P(RDFSimpleTests, StandardDeviationEmpty)
...
+
+
+TEST(RDFSimpleTests, KahanSum)
+{
+   ROOT::RDataFrame d(20);
+   auto dd = d.Define("x", "(rdfentry_ %2 == 0) ? 0.00000001 : 100000000.");
+   EXPECT_EQ(*dd.Sum<double>({"x"}), 1000000000.00000011920929);
```
Comment (eguiraud) [inline tree/dataframe/test/dataframe_simple.cxx]: as discussed I am not a big fan of this particular test setup as the errors are at the 17th significant digit.

it would be better to have a test for `float`s and a test for `double`s that show numerical instability at digits well within their expected precision.

also testing `double`s for equality is very tricky (on different architectures we might get slightly different results). it would be better to use `EXPECT_DOUBLE_EQ` or `EXPECT_NEAR`, see also https://google.github.io/googletest/reference/assertions.html#floating-point

File: tree/dataframe/test/dataframe_simple.cxx [hunk 4]
```diff
@@ -726,12 +726,21 @@ TEST_P(RDFSimpleTests, StandardDeviationEmpty)
...
+/*
 TEST(RDFSimpleTests, SumOfStrings)
 {
    auto df = RDataFrame(2).Define("str", []() -> std::string { return "bla"; });
    EXPECT_EQ(*df.Sum<std::string>("str"), "blabla");
 }
+*/
+
+
+TEST(RDFSimpleTests, KahanSum)
```
Comment (eguiraud) [inline tree/dataframe/test/dataframe_simple.cxx:738]: in this commit and the next, using `TEST_P` instead of `TEST` would generate a single-thread _and_ a multi-thread version of the test (because of gtest's test parameterization + the last few lines in this file)

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx [hunk 5]
```diff
@@ -1186,7 +1186,7 @@ public:
          compensation = (t - sum) - y;
          sum = t;
       }
-   *fResultSum = sum;
+   *fResultSum += sum;
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1189]: could we simplify this by removing the local `sum` variable and directly accumulate into `fResultSum` in the loop above?

**Review by eguiraud — CHANGES_REQUESTED** [review]
Thank you @j-mathe , overall it looks like what we want. I left a bunch of minor comments/suggestions, and before merging we need to

- [ ] add a commit that applies clang-format fixes (e.g. with `git clang-format --diff master HEAD | git apply`) (Julia)
- [ ] figure out what to do with the sum of strings -- add a specialization of `SumHelper` to still support it or just drop it (Enrico)
- [x] check why MT test with floats performs poorly

**Comment by eguiraud** [comment]
@j-mathe please fix the compilation warnings reported by the CI when you have a second :grimacing:

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx [hunk 6]
```diff
@@ -1187,6 +1239,7 @@ class R__CLING_PTRCHECK(off) MeanHelper : public RActionImpl<MeanHelper> {
    std::vector<ULong64_t> fCounts;
    std::vector<double> fSums;
    std::vector<double> fPartialMeans;
+   std::vector<double> fCompensations;
```
Comment (ikabadzhov) [inline tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1242]: (Why) do we need a global variable `fCompensations`? Can we pass it as an extra (local) argument to the functions that require it?

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx [hunk 7]
```diff
@@ -1142,26 +1173,47 @@ public:
...
-         *fResultSum += m;
+      ResultType sum(NeutralElement(ResultType{}, -1));
+      ResultType compensation(NeutralElement(ResultType{}, -1));
+      ResultType y(NeutralElement(ResultType{}, -1));
+      ResultType t(NeutralElement(ResultType{}, -1));
+      for (auto &m : fSums) {
+         // Kahan Sum:
+         y = m - compensation;
+         t = sum + y;
+         compensation = (t - sum) - y;
```
Comment (ikabadzhov) [inline tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1213]: shouldn't this take a vector of compensations?

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx [hunk 8]
```diff
@@ -1120,8 +1149,10 @@ public:
 
 template <typename ResultType>
 class R__CLING_PTRCHECK(off) SumHelper : public RActionImpl<SumHelper<ResultType>> {
-   const std::shared_ptr<ResultType> fResultSum;
+   std::shared_ptr<ResultType> fResultSum;
```
Comment (ikabadzhov) [inline tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1152]: why did you remove the const-ness?

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx [hunk 9]
```diff
@@ -1120,8 +1149,10 @@ public:
...
    Results<ResultType> fSums;
+   Results<ResultType> fCompensations;
+   unsigned int fNSlots;
```
Comment (ikabadzhov) [inline tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1155]: why do we need to save the fNSlots here?

**Comment by eguiraud** [comment]
@phsft-bot build please

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx (hunk 6 above)
Comment (j-mathe) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1242]: Yes, it should be possible to pass it as an extra argument in void Exec(unsigned int slot, ResultType x), for instance.
Would that be better?

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx (hunk 7 above)
Comment (j-mathe) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1213]: No, because this is just the Kahan sum of the final contributions which are already sums and not vectors.

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx (hunk 8 above)
Comment (j-mathe) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1152]: Because there was no const-ness in the Kahan sum tutorial where the Kahan class was defined. I think I can apply the const again, if that's better?

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx (hunk 9 above)
Comment (j-mathe) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1155]: Ah, thank you, that has become obsolete and can be removed, I believe.

**Comment by eguiraud** [comment]
Hi @j-mathe , I think you used `clang-format filename1 filename2 ...` instead of `git clang-format --diff ...`, right? because the clang-format commit introduces many formatting changes that are unrelated to the PR and that we should not apply here.

It also makes it a bit harder to review the PR.

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx (hunk 8 above)
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx]: Here it's not the number that is marked const, it's the pointer -- this is a constant shared pointer to a non-constant number.

Note that in general const data members are discouraged: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-constref -- but in this case it's also a valid approach to leave the code as you found it :)

File: tree/dataframe/src/RDFActionHelpers.cxx [hunk 10]
```diff
@@ -133,21 +133,33 @@ BufferedFillHelper::Exec(unsigned int, const std::vector<unsigned int> &, const
 // template void MaxHelper::Exec(unsigned int, const std::vector<unsigned int> &);
 
 MeanHelper::MeanHelper(const std::shared_ptr<double> &meanVPtr, const unsigned int nSlots)
-   : fResultMean(meanVPtr), fCounts(nSlots, 0), fSums(nSlots, 0), fPartialMeans(nSlots)
+   : fResultMean(meanVPtr), fCounts(nSlots, 0), fSums(nSlots, 0), fPartialMeans(nSlots), fCompensations(nSlots)
```
Comment (ikabadzhov) [inline tree/dataframe/src/RDFActionHelpers.cxx:136]: ooooo problem might be here! `fCompensations` are not initialized!

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx (hunk 6 above)
Comment (ikabadzhov) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1210]: okay, I think it is okay to remain like that, but see my latest comment. There was a place which did not initialize the elements of fCompensations!

File: tree/dataframe/src/RDFActionHelpers.cxx (hunk 10 above)
Comment (eguiraud) [reply tree/dataframe/src/RDFActionHelpers.cxx:136]: that's not right, `std::vector` default-initializes its contents: https://en.cppreference.com/w/cpp/container/vector/vector (constructor number 4) and https://godbolt.org/z/ePjshzMnh .

there have been no failures in the last few builds, I think things are okay.

File: tree/dataframe/test/dataframe_simple.cxx [hunk 11]
```diff
@@ -727,12 +727,49 @@ TEST_P(RDFSimpleTests, StandardDeviationEmpty)
    EXPECT_DOUBLE_EQ(*stdDev, 0);
 }
 
+/*
```
Comment (eguiraud) [inline tree/dataframe/test/dataframe_simple.cxx:730]:
```suggestion
/* Since the introduction of Kahan sums in Sum and Mean, we require that T in `Sum<T>` has an `operator-`
```

**Review by eguiraud — APPROVED** [review]
LGTM, we should just comment why SumOfStrings is disabled.

In case anyone ever complains that we broke that use case we can add a different code path for `Sum`s of types without an `operator-`.
